### PR TITLE
BFD-2126: Add ability for pipeline to read and write to a new location for synthetic data

### PR DESCRIPTION
**Pull request by lsmitchell**
**JIRA Ticket:**
[BFD-2126](https://jira.cms.gov/browse/BFD-2126)

**User Story or Bug Summary:**

As a BFD engineer, I would like to have a second location that the ETL loads from, so that I can leave the Incoming folder uncluttered when I have hundreds of weeks future data folders created during synthetic data generation.

---

### What Does This PR Do?

Updates a number of places needed to read loads from a new location, Synthetic/Incoming, and write the finished files from this location into Synthetic/Done. This will allow us to keep synthetic "future" loads in a storage location that does not clutter up the standard pipeline Incoming/Done folders.

- Add two new S3 prefix locations that the pipeline uses, Synthetic/Incoming and Synthetic/Done
    - Synthetic/Incoming is treated the same as Incoming for all intents and purposes, this is simply a second load location for organization and separation concerns
    - Synthetic/Done will be used to store any loads detected from within Synthetic/Incoming to keep the regular Done folder uncluttered for ease of debugging real loads in production
- In various places where we check the manifests/files, there is a switch that will verify if we're loading from Incoming or Synthetic/Incoming, so that we can grab the files from the right location, and move them to the corresponding location
    - Since this info on the manifest parent key isnt readily available in most places, we check the key by checking if the manifest (or one of its files) exists in one key (Incoming) or the other (Synthetic/Incoming); it should only ever exist in one place so this check should be sufficient
- Small refactoring where needed to aid in readability/reusability on areas touched
- Adds IT tests to check that items can be read from Synthetic/Incoming and are written out to Synthetic/Done

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.
* If you know of any documentation pointing to the load locations of pipeline, let me know so I can update them

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by christrotteradhoc — COMMENTED** [review]
Nice job finding all of the parts that need to change! I think it works but I've got a few suggestions that I believe could make it tidier.

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/CcwRifLoadJob.java [hunk 1]
```diff
@@ -285,8 +318,32 @@ private boolean dataSetIsAvailable(DataSetManifest manifest) {
...
+      // files
+      return false;
+    } else if (!s3TaskManager
+        .getS3Client()
+        .doesObjectExist(options.getS3BucketName(), prefix + sampleItemLocation)) {
+      prefix = syntheticPrefix;
+    }
+
+    // Value must be final for lambda expression using them to be happy
+    final String dataSetKeyPrefix = prefix;
```
[the remark this answers is not in the document]
Comment (lsmitchell) [reply apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/CcwRifLoadJob.java]: Yeah I was torn between that and this when I was thinking through this initially. Timeboxed playing around with that way after your comment and I think it is easier to read/understand, so I refactored it to be set/stored in the manifest data which simplifies the other places.

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/CcwRifLoadJob.java [hunk 2]
```diff
@@ -102,9 +115,29 @@ public final class CcwRifLoadJob implements PipelineJob<NullPipelineJobArguments
   public static final Pattern REGEX_PENDING_MANIFEST =
       Pattern.compile("^" + S3_PREFIX_PENDING_DATA_SETS + "\\/(.*)\\/([0-9]+)_manifest\\.xml$");
 
+  /**
+   * A regex for {@link DataSetManifest} keys in S3 for synthetic files. Provides capturing groups
+   * for the {@link DataSetManifestId} fields.
+   */
+  public static final Pattern REGEX_PENDING_MANIFEST_SYNTHETIC =
+      Pattern.compile(
+          "^" + S3_PREFIX_PENDING_SYNTHETIC_DATA_SETS + "\\/(.*)\\/([0-9]+)_manifest\\.xml$");
```
[the remark this answers is not in the document]
Comment (lsmitchell) [reply apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/CcwRifLoadJob.java]: True, adjusted it to just be one regex.

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/CcwRifLoadJob.java [hunk 3]
```diff
@@ -102,9 +115,29 @@ public final class CcwRifLoadJob implements PipelineJob<NullPipelineJobArguments
...
       Pattern.compile("^" + S3_PREFIX_COMPLETED_DATA_SETS + "\\/(.*)\\/([0-9]+)_manifest\\.xml$");
 
+  /**
+   * A regex that can be used for checking for a manifest in the {@link
+   * #S3_PREFIX_COMPLETED_SYNTHETIC_DATA_SETS} location.
+   */
+  public static final Pattern REGEX_COMPLETED_MANIFEST_SYNTHETIC =
+      Pattern.compile(
+          "^" + S3_PREFIX_COMPLETED_SYNTHETIC_DATA_SETS + "\\/(.*)\\/([0-9]+)_manifest\\.xml$");
+
```
[the remark this answers is not in the document]
Comment (lsmitchell) [reply apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/CcwRifLoadJob.java]: Adjusted!

**Review by christrotteradhoc — APPROVED** [review]

**Review by mjburling — APPROVED** [review]

**Pull request merged**